### PR TITLE
All gatherwith grads

### LINKED ISSUE #4725: Gather function in Lightning module with gradients
## 🚀 Feature

Add a gather function with gradients in LightningModule which can be implemented with the help of:

https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/models/self_supervised/simclr/simclr_module.py#L25

### Motivation

The default ```all_gather``` in pytorch breaks the gradient graph as it does not propagate any gradient backwards.

### Pitch

Allow users to use self.gather in LightningModule.

```
x = self.gather(x)
```
As suggested by @williamFalcon, this can be implemented like:

```
def gather(self):
    self.trainer.gather():


Trainer.gather(self, x):
    x = self.gather.accelerator(x)
```

All accelerators except CPU should implement this. 

The gradients can be synced with:

```
x = self.gather(x, sync_grads=True)
```

@edenlightning @tchaton @teddykoker

### PR DESCRIPTION
**Pull request by ananyahjha93**
## What does this PR do?

Fixes #4725.

Adds all_gather method to Lightning module with gradients.

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/distributed.py [hunk 1]
```diff
@@ -155,3 +159,61 @@ def sync_ddp(
         result = result / torch.distributed.get_world_size(group)
 
     return result
+
+
+class AllGatherGrad(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, tensor, group=group.WORLD):
+        ctx.batch_size = tensor.shape[0]
```
Comment (tchaton) [inline pytorch_lightning/utilities/distributed.py:167]: Why saving the batch_size ?

File: pytorch_lightning/utilities/distributed.py [hunk 2]
```diff
@@ -155,3 +159,61 @@ def sync_ddp(
...
+        ctx.group = group
+
+        gathered_tensor = [
```
Comment (tchaton) [inline pytorch_lightning/utilities/distributed.py:170]: Could you extend this to support tuple of tensors ?

**Review by tchaton — COMMENTED** [review]
Looks great ! Can you extend support to tuple of tensors ? Ideally, we will also support dictionaries.

File: pytorch_lightning/utilities/distributed.py (hunk 2 above)
Comment (ananyahjha93) [reply pytorch_lightning/utilities/distributed.py:170]: tuple of tensors how? Like gather a tuple of tensors across processes?

File: pytorch_lightning/utilities/distributed.py (hunk 1 above)
Comment (ananyahjha93) [reply pytorch_lightning/utilities/distributed.py]: removed

**Comment by ananyahjha93** [comment]
@tchaton so right now the idea is to just extend the default all_gather with gradients in the LightningModule itself. Needs to be included in the 1.1. release. We can add support remaining features in the next release after a discussion.

File: pytorch_lightning/accelerators/accelerator.py [hunk 3]
```diff
@@ -175,6 +175,20 @@ def sync_tensor(self,
...
+
+        Args:
+            tensor: tensor of shape (batch, ...)
+            group: the process group to gather results from. Defaults to all processes (world)
+            sync_grads: flag that allows users to synchronize gradients for all_gather op
+
+        Return:
+            A tensor of shape (world_size, batch, ...)
+        """
+        raise NotImplementedError()
```
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:190]:
```suggestion
        return all_gather_ddp_if_available(tensor, group=group, sync_grads=sync_grads)
```
it seems to be everywhere the same, can we rather implement it here and drop in child if needed

Comment (SeanNaren) [reply pytorch_lightning/accelerators/accelerator.py:190]: I would suggest against this, because we shouldn't assume that it can be directly implemented in every accelerator, a lot of duplication I know but it's best since theres no accelerator subclass

File: pytorch_lightning/utilities/distributed.py [hunk 4]
```diff
@@ -155,3 +159,60 @@ def sync_ddp(
...
+
+    Args:
+        tensor: tensor of shape (batch, ...)
+        group: the process group to gather results from. Defaults to all processes (world)
+        sync_grads: flag that allows users to synchronize gradients for all_gather op
+
+    Return:
+        A tensor of shape (world_size, batch, ...)
+    """
+    if torch.distributed.is_available() and torch.distributed.is_initialized():
```
Comment (SeanNaren) [inline pytorch_lightning/utilities/distributed.py:206]: Did you require this check? I feel like this should all gather or crash.

**Review by SeanNaren — APPROVED** [review]
Small nits but looks good!

File: pytorch_lightning/utilities/distributed.py (hunk 4 above)
Comment (justusschock) [reply pytorch_lightning/utilities/distributed.py:206]: I think so as well, we should only check this where we use this function...

File: pytorch_lightning/utilities/distributed.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/utilities/distributed.py:169]: @tchaton I think we shouldn't do this. IMO This should only be working on a single tensor and if we want to support tuples, we should just call this function onto the tuples' elements (e.g. with `apply_to_collection`), but we really shouldn't include this in that function here!

File: pytorch_lightning/utilities/distributed.py [hunk 5]
```diff
@@ -155,3 +159,60 @@ def sync_ddp(
...
+        if sync_grads:
+            return AllGatherGrad.apply(tensor, group)
```
Comment (justusschock) [inline pytorch_lightning/utilities/distributed.py:208]: can't we use this function always and if `sync_grads` is False, just use it with` torch.no_grad`?

That way we have at least a unified behaviour (especially when `AllGatherGrad` will return tuples.

**Review by justusschock — APPROVED** [review]
LGTM. Just some minor things.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/utilities/distributed.py (hunk 4 above)
Comment (ananyahjha93) [reply pytorch_lightning/utilities/distributed.py:206]: I thought the idea of if_available was to make this check? Similar pattern followed in https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/utilities/distributed.py#L106

File: pytorch_lightning/accelerators/accelerator.py (hunk 3 above)
Comment (ananyahjha93) [reply pytorch_lightning/accelerators/accelerator.py:190]: @Borda @SeanNaren I agree with Sean, majority of accelerators are some variant of DDP, that is why there is quite a bit of duplication. However, the same thing won't work for TPU and horovod

File: pytorch_lightning/utilities/distributed.py (hunk 5 above)
Comment (ananyahjha93) [reply pytorch_lightning/utilities/distributed.py:208]: updated

File: pytorch_lightning/utilities/distributed.py (hunk 4 above)
Comment (SeanNaren) [reply pytorch_lightning/utilities/distributed.py:206]: Ah didn't know that this is based on the same pattern, then I think it's fine :)

File: pytorch_lightning/utilities/distributed.py [hunk 6]
```diff
@@ -155,3 +159,54 @@ def sync_ddp(
...
+        torch.distributed.all_gather(gathered_tensor, tensor, group=group)
+        gathered_tensor = torch.stack(gathered_tensor, dim=0)
+
+        return gathered_tensor
+
+    @staticmethod
+    def backward(ctx, *grad_output):
+        grad_output = torch.cat(grad_output)
+
+        torch.distributed.all_reduce(
```
Comment (ajtulloch) [inline pytorch_lightning/utilities/distributed.py:182]: Can you just use `reduce_scatter`? Using a full `all_reduce` is unnecessary.
